website/docs/BestPractices.md

```diff
@@ -0,0 +1,186 @@
+
+It isn't always possible to do this but using CSS pseudo-classes like [:nth-child](https://developer.mozilla.org/en-US/docs/Web/CSS/:nth-child) you can match elements based on the indexes of the elements in the child list of their parents.
+
+Queries all table rows.
+
+```js
+// 👎
+await $$('table tr')[15];
+```
+
+Queries a single table row.
+
+```js
+// 👍
+await $('table tr:nth-child(15)');
+```
+
+## Use the built-in assertions
+
+Don't use manual assertions that do not automatically wait for the results to match as this will cause for flaky tests.
+
+```js
+// 👎
+expect(await button.isDisplayed()).toBe(true);
+```
+
+By using the built-in assertions WebdriverIO will automatically wait for the actual result to match the expected result resulting in resilient tests.
+
+```js
+// 👍
+await expect(button).toBeDisplayed();
+```
+
+## Don't overuse commands and assertions
+
+When using expect.toBeDisplayed you implicitly also wait for the element to exist. There isn't a need to use the waitForXXX commands when you already have an assertion doing the same thing.
+
+```js
+// 👎
+await button.waitForExist();
+await expect(button).toBeDisplayed();
+
+// 👎
+await button.waitForDisplayed();
+await expect(button).toBeDisplayed();
+
+// 👍
+await expect(button).toBeDisplayed();
+```
+
+No need to wait for an element to exist or be displayed when interacting or when asserting something like it's text unless the element can explicitly be invisible (opacity: 0; for example) or can explicitly be disabled (disabled attribute for example) in which case waiting for the element to be displayed makes sense.
+
+```js
+// 👎
+await expect(button).toBeExisting();
+await expect(button).toHaveText('Submit');
+
+// 👎
+await expect(button).toBeDisplayed();
+await expect(button).toHaveText('Submit');
+
+// 👎
+await expect(button).toBeDisplayed();
+await expect(button).click();
+```
+
+```js
+// 👍
+await expect(button).click();
+
+// 👍
+await expect(button).toHaveText('Submit');
+```
+
+## Lint your code
+
+Using eslint to lint your code you can potentionally catch errors early, use our [linting rules](https://www.npmjs.com/package/eslint-plugin-wdio) to make sure that some of the best practices are always applied.
+
+## Don't pause
+
+It can be tempting to use the pause command but using this is a bad idea as it isn't resilient and will only cause for flaky tests in the long run.
+
+```js
+// 👎
+await nameInput.setValue('Bob');
+await browser.pause(200); // wait for submit button to enable
+await submitFormButton.click();
+
+// 👍
+await nameInput.setValue('Bob');
+await submitFormButton.waitForEnabled();
+await submitFormButton.click();
+```
+
+## Async loops
+
+When you have some asynchronous code that you want to repeat, it is important to know that not all loops can do this.
+
+The following will not click the elements in order.
+
+```js
+// 👎
+elements.forEach(async (element) => {
+    await element.click();
+});
```

Comment:
Actually our interface allows this and I am using it regularly e.g. to fetch a set of elements and get their text I would do:

```ts
console.log(await $$('h1').map((h1) => h1.getText()))
```

Something we should point out here is that your example above would run all clicks in parallel while you actually want to run them sequentially. We should point that out. Using array methods on elements should __not__ be discouraged.

Reply:
> Actually our interface allows this and I am using it regularly e.g. to fetch a set of elements and get their text I would do:
> 
> ```ts
> console.log(await $$('h1').map((h1) => h1.getText()))
> ```
> 
> Something we should point out here is that your example above would run all clicks in parallel while you actually want to run them sequentially. We should point that out. Using array methods on elements should **not** be discouraged.

You are correct with that example, however let me show you what I wanted to make our users aware of.

Given we have a table with table rows that have some text in order from 0 to 19 as show here:
<img width="109" alt="Screenshot 2023-09-02 at 22 24 50" src="https://github.com/webdriverio/webdriverio/assets/15839059/398ca07f-98c3-4b19-bb7c-20380b0896ae">

When I use forEach like shown in the screenshot, you'll see that the text is returned in a random order:
<img width="1146" alt="Screenshot 2023-09-02 at 22 25 15" src="https://github.com/webdriverio/webdriverio/assets/15839059/e010c178-0999-4f9e-9811-48216531fc91">

And when using the for...of loop it returns the text as expected:
<img width="1144" alt="Screenshot 2023-09-02 at 22 26 56" src="https://github.com/webdriverio/webdriverio/assets/15839059/2b48ba26-fab3-4c4a-9c3d-07fae30e368f">

I think we are both right in this case, what do you think?

Reply:
I have updated the example that might make it more clear but I'm unsure at this point if it is actually any clearer or not